Archive/Imo/Imo2024Q3.lean

```diff
@@ -0,0 +1,961 @@
+
+/-!
+# IMO 2024 Q3
+
+Let $a_1, a_2, a_3, \ldots$ be an infinite sequence of positive integers, and let $N$ be a
+positive integer. Suppose that, for each $n > N$, $a_n$ is equal to the number of times
+$a_{n-1}$ appears in the list $a_1, a_2, \ldots, a_{n-1}$.
+
+Prove that at least one of the sequences $a_1, a_3, a_5, \ldots$ and $a_2, a_4, a_6, \ldots$
+is eventually periodic.
+
+We follow Solution 1 from the
+[official solutions](https://www.imo2024.uk/s/IMO2024-solutions-updated.pdf). We show that
+the positive integers up to some positive integer $k$ (referred to as small numbers) all appear
+infinitely often in the given sequence while all larger positive integers appear only finitely
+often and then that the sequence eventually alternates between small numbers and larger integers.
+A further detailed analysis of the eventual behavior of the sequence ends up showing that the
+sequence of small numbers is eventually periodic with period at most $k$ (in fact exactly $k$).
+-/
+
+
+open scoped Finset
+
+namespace Imo2024Q3
+
+/-- The condition of the problem. Following usual Lean conventions, this is expressed with
+indices starting from 0, rather than from 1 as in the informal statment (but `N` remains as the
+index of the last term for which `a n` is not defined in terms of previous terms). -/
+def Condition (a : ℕ → ℕ) (N : ℕ) : Prop :=
+  (∀ i, 0 < a i) ∧ ∀ n, N < n → a n = #{i ∈ (Finset.range n) | a i = a (n - 1)}
+
+/-- The property of a sequence being eventually periodic. -/
+def EventuallyPeriodic (b : ℕ → ℕ) : Prop := ∃ p M, 0 < p ∧ ∀ m, M ≤ m → b (m + p) = b m
+
+/-! ### Definitions and lemmas about the sequence that do not actually need the condition of
+the problem -/
+
+/-- A number greater than any of the initial terms `a 0` through `a N` (otherwise arbitrary). -/
+def M (a : ℕ → ℕ) (N : ℕ) : ℕ := (Finset.range (N + 1)).sup a + 1
+
+lemma M_pos (a : ℕ → ℕ) (N : ℕ) : 0 < M a N :=
+  Nat.add_one_pos _
+
+lemma one_le_M (a : ℕ → ℕ) (N : ℕ) : 1 ≤ M a N :=
+  Nat.lt_iff_add_one_le.1 (M_pos a N)
+
+lemma apply_lt_M_of_le_N (a : ℕ → ℕ) {N i : ℕ} (h : i ≤ N) : a i < M a N :=
+  Nat.lt_add_one_iff.2 (Finset.le_sup (Finset.mem_range_succ_iff.2 h))
+
+lemma N_lt_of_M_le_apply {a : ℕ → ℕ} {N i : ℕ} (h : M a N ≤ a i) : N < i := by
+  by_contra! hi
+  exact Nat.not_succ_le_self _ (h.trans (Finset.le_sup (Finset.mem_range_succ_iff.2 hi)))
+
+lemma ne_zero_of_M_le_apply {a : ℕ → ℕ} {N i : ℕ} (h : M a N ≤ a i) : i ≠ 0 :=
+  Nat.not_eq_zero_of_lt (N_lt_of_M_le_apply h)
+
+lemma apply_lt_of_M_le_apply {a : ℕ → ℕ} {N i j : ℕ} (hi : M a N ≤ a i) (hj : j ≤ N) :
+    a j < a i :=
+  (apply_lt_M_of_le_N a hj).trans_le hi
+
+lemma apply_ne_of_M_le_apply {a : ℕ → ℕ} {N i j : ℕ} (hi : M a N ≤ a i) (hj : j ≤ N) :
+    a j ≠ a i :=
+  (apply_lt_of_M_le_apply hi hj).ne
+
+lemma toFinset_card_pos {a : ℕ → ℕ} {i : ℕ} (hf : {j | a j = a i}.Finite) : 0 < #hf.toFinset :=
+  Finset.card_pos.mpr ((Set.Finite.toFinset_nonempty _).mpr ⟨i, rfl⟩)
```

Comment:
This can be generalized to `{y | y = x}`, with `x : X` in any type.

Reply:
I don't think `{y | y = x}` is a generalization of `{j | a j = a i}`, but also the natural general statements are the individual pre-existing mathlib lemmas used in the proof. This lemma is just a particular combination of those mathlib lemmas that proves convenient in the context of this problem; I don't really see this particular combination (even with more general types), from very many possible such combinations, as being of use for mathlib proper.

Reply:
Aah, you are right. I was too quick with commenting. Sorry.